x/reports/types/models.go

```diff
@@ -0,0 +1,187 @@
+package types
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/gogo/protobuf/proto"
+)
+
+// ParseReportID parses the given value as a report id, returning an error if it's invalid
+func ParseReportID(value string) (uint64, error) {
+	if value == "" {
+		return 0, nil
+	}
+
+	reportID, err := strconv.ParseUint(value, 10, 64)
+	if err != nil {
+		return 0, fmt.Errorf("invalid report id: %s", err)
+	}
+	return reportID, nil
+}
+
+// NewReport returns a new Report instance
+func NewReport(
+	subspaceID uint64,
+	id uint64,
+	reasonID uint32,
+	message string,
+	data ReportData,
+	reporter string,
+	creationDate time.Time,
+) Report {
+	dataAny, err := codectypes.NewAnyWithValue(data)
+	if err != nil {
+		panic("failed to pack data to any type")
+	}
+
+	return Report{
+		SubspaceID:   subspaceID,
+		ID:           id,
+		ReasonID:     reasonID,
+		Message:      message,
+		Data:         dataAny,
+		Reporter:     reporter,
+		CreationDate: creationDate,
+	}
+}
+
+// Validate implements fmt.Validator
+func (r Report) Validate() error {
+	if r.SubspaceID == 0 {
+		return fmt.Errorf("invalid subspace id: %d", r.Size())
+	}
+
+	if r.ID == 0 {
+		return fmt.Errorf("invalid report id: %d", r.ID)
+	}
+
+	if r.ReasonID == 0 {
+		return fmt.Errorf("invalid reason id: %d", r.ReasonID)
+	}
+
+	_, err := sdk.AccAddressFromBech32(r.Reporter)
+	if err != nil {
+		return fmt.Errorf("invalid reporter address: %s", err)
+	}
+
+	err = r.Data.GetCachedValue().(ReportData).Validate()
+	if err != nil {
+		return err
+	}
+
+	if r.CreationDate.IsZero() {
+		return fmt.Errorf("invalid report creation date: %s", r.CreationDate)
+	}
+
+	return nil
+}
+
+// UnpackInterfaces implements codectypes.UnpackInterfacesMessage
+func (r *Report) UnpackInterfaces(unpacker codectypes.AnyUnpacker) error {
+	var data ReportData
+	return unpacker.UnpackAny(r.Data, &data)
+}
+
+// --------------------------------------------------------------------------------------------------------------------
+
+// ReportData represents a generic report data
+type ReportData interface {
```

Comment:
I would name it `ReportTarget`.

Reply:
Renamed everything from _report data_ to _report target_